From 7b864d8e52606ce71436e697f2f592472ddc5465 Mon Sep 17 00:00:00 2001
From: Yordan Dimitrov <y.dimitrov.14@gmail.com>
Date: Mon, 4 Jun 2018 17:29:50 +0300
Subject: [PATCH 1/3] Changing LED colour gradient according to temperature
 sliders

v1
- the colour of the LED reflects the temperature of both sliders
  at the same time
- colour gradient starts from blue and ends with red:
  for low and high temperatures, respectively.

v2
- removed all A/C code
- paths are read from a predefined variables

Change-Id: I989d6779c0a72b9def74bc9bcf41ae01d368ba8a
Signed-off-by: Yordan Dimitrov <y.dimitrov.14@gmail.com>
---
 app/HVAC.qml                |   2 +
 app/api/Binding.qml         |  10 +
 binding/hvac-demo-binding.c | 857 ++++++++++++++++++++++++++++----------------
 3 files changed, 570 insertions(+), 299 deletions(-)

diff --git a/app/HVAC.qml b/app/HVAC.qml
index b1925dd..be16e3f 100644
--- a/app/HVAC.qml
+++ b/app/HVAC.qml
@@ -138,6 +138,8 @@ ApplicationWindow {
                         binding.rightTemperature = degree
                     }
                 }
+
+
             }
         }
 
diff --git a/app/api/Binding.qml b/app/api/Binding.qml
index 25388df..7d996ee 100644
--- a/app/api/Binding.qml
+++ b/app/api/Binding.qml
@@ -29,6 +29,7 @@ WebSocket {
     property real rightTemperature: 21.0
     property string language: "en_US"
 
+
     property Connections c : Connections {
         target: root
         onFanSpeedChanged: {
@@ -40,11 +41,20 @@ WebSocket {
             var json = [MessageId.call, '9999', 'hvac/set', {'LeftTemperature': leftTemperature}]
             console.debug(JSON.stringify(json))
             sendTextMessage(JSON.stringify(json))
+
+            var json1 = [MessageId.call, '9999', 'hvac/temp_left_zone_led', {'LeftLed': leftTemperature}]
+            console.debug(JSON.stringify(json1))
+            sendTextMessage(JSON.stringify(json1))
+               
         }
         onRightTemperatureChanged: {
             var json = [MessageId.call, '9999', 'hvac/set', {'RightTemperature': rightTemperature}]
             console.debug(JSON.stringify(json))
             sendTextMessage(JSON.stringify(json))
+
+            var json1 = [MessageId.call, '9999', 'hvac/temp_right_zone_led', {'RightLed': rightTemperature}]
+            console.debug(JSON.stringify(json1))
+            sendTextMessage(JSON.stringify(json1))
         }
         onLanguageChanged: {
             var json = [MessageId.call, '9999', 'hvac/set', {'Language': language}]
diff --git a/binding/hvac-demo-binding.c b/binding/hvac-demo-binding.c
index 8565b56..4ac8d6e 100644
--- a/binding/hvac-demo-binding.c
+++ b/binding/hvac-demo-binding.c
@@ -19,6 +19,7 @@
  */
 #define _GNU_SOURCE
 
+#include <stdio.h>
 #include <string.h>
 #include <stdbool.h>
 #include <unistd.h>
@@ -32,6 +33,9 @@
 #include <json-c/json.h>
 
 #define AFB_BINDING_VERSION 2
+#define RED "/sys/class/leds/blinkm-3-9-red/brightness"
+#define BLUE "/sys/class/leds/blinkm-3-9-blue/brightness"
+#define GREEN "/sys/class/leds/blinkm-3-9-green/brightness"
 #include <afb/afb-binding.h>
 
 #define CAN_DEV "vcan0"
@@ -45,11 +49,11 @@ static struct afb_event event;
 
 /*****************************************************************************************/
 /*****************************************************************************************/
-/**											**/
-/**											**/
-/**	   SECTION: UTILITY FUNCTIONS							**/
-/**											**/
-/**											**/
+/**                                         **/
+/**                                         **/
+/**    SECTION: UTILITY FUNCTIONS                           **/
+/**                                         **/
+/**                                         **/
 /*****************************************************************************************/
 /*****************************************************************************************/
 
@@ -63,186 +67,425 @@ static struct afb_event event;
  */
 static int retry( int(*func)())
 {
-	int i;
-
-	for (i=0;i<4;i++)
-	{
-		if ( (*func)() >= 0)
-		{
-			return 0;
-		}
-		usleep(100000);
-	}
-	return -1;
+    int i;
+
+    for (i=0;i<4;i++)
+    {
+        if ( (*func)() >= 0)
+        {
+            return 0;
+        }
+        usleep(100000);
+    }
+    return -1;
 }
 
 /*****************************************************************************************/
 /*****************************************************************************************/
-/**											**/
-/**											**/
-/**	   SECTION: HANDLE CAN DEVICE							**/
-/**											**/
-/**											**/
+/**                                         **/
+/**                                         **/
+/**    SECTION: HANDLE CAN DEVICE                           **/
+/**                                         **/
+/**                                         **/
 /*****************************************************************************************/
 /*****************************************************************************************/
 
 // Initialize CAN hvac array that will be sent trough the socket
 static struct {
-	const char *name;
-	uint8_t value;
+    const char *name;
+    uint8_t value;
 } hvac_values[] = {
-	{ "LeftTemperature", 21 },
-	{ "RightTemperature", 21 },
-	{ "Temperature", 21 },
-	{ "FanSpeed", 0 }
+    { "LeftTemperature", 21 },
+    { "RightTemperature", 21 },
+    { "Temperature", 21 },
+    { "FanSpeed", 0 },
+    { "ACEnabled", 0 },
+    { "LeftLed", 15 },
+    { "RightLed", 15 }
+};
+
+// Holds RGB combinations for each temperature
+static struct {
+    const int temperature;
+    const int rgb[3];
+} degree_colours[] = {
+    {15, {0, 0, 229} },
+    {16, {22, 0, 204} },
+    {17, {34, 0, 189} },
+    {18, {46, 0, 175} },
+    {19, {58, 0, 186} },
+    {20, {70, 0, 146} },
+    {21, {82, 0, 131} },
+    {22, {104, 0, 116} },
+    {23, {116, 0, 102} },
+    {24, {128, 0, 87} },
+    {25, {140, 0, 73} },
+    {26, {152, 0, 58} },
+    {27, {164, 0, 43} },
+    {28, {176, 0, 29} },
+    {29, {188, 0, 14} },
+    {30, {201, 0, 5} }
 };
 
 struct can_handler {
-	int socket;
-	bool simulation;
-	char *send_msg;
-	struct sockaddr_can txAddress;
+    int socket;
+    bool simulation;
+    char *send_msg;
+    struct sockaddr_can txAddress;
 };
 
 static struct can_handler can_handler = { .socket = -1, .simulation = false, .send_msg = "SENDING CAN FRAME"};
 
 static int open_can_dev_helper()
 {
-	struct ifreq ifr;
-
-	AFB_DEBUG("CAN Handler socket : %d", can_handler.socket);
-	close(can_handler.socket);
-
-	can_handler.socket = socket(PF_CAN, SOCK_RAW, CAN_RAW);
-	if (can_handler.socket < 0)
-	{
-		AFB_ERROR("socket could not be created");
-	}
-	else
-	{
-		// Attempts to open a socket to CAN bus
-		strcpy(ifr.ifr_name, CAN_DEV);
-		if(ioctl(can_handler.socket, SIOCGIFINDEX, &ifr) < 0)
-		{
-			AFB_ERROR("ioctl failed");
-		}
-		else
-		{
-			can_handler.txAddress.can_family = AF_CAN;
-			can_handler.txAddress.can_ifindex = ifr.ifr_ifindex;
-
-			// And bind it to txAddress
-			if (bind(can_handler.socket, (struct sockaddr *)&can_handler.txAddress, sizeof(can_handler.txAddress)) < 0)
-			{
-				AFB_ERROR("bind failed");
-			}
-			else {
-				return 0;
-			}
-		}
-		close(can_handler.socket);
-		can_handler.socket = -1;
-	}
-	return -1;
+    struct ifreq ifr;
+    
+    AFB_DEBUG("CAN Handler socket : %d", can_handler.socket);
+    close(can_handler.socket);
+
+    can_handler.socket = socket(PF_CAN, SOCK_RAW, CAN_RAW);
+    if (can_handler.socket < 0)
+    {
+        AFB_ERROR("ld not created");
+    }
+    else
+    {
+        // Attempts to open a socket to CAN bus
+        strcpy(ifr.ifr_name, CAN_DEV);
+        if(ioctl(can_handler.socket, SIOCGIFINDEX, &ifr) < 0)
+        {
+            AFB_ERROR("ed");
+        }
+        else
+        {
+            can_handler.txAddress.can_family = AF_CAN;
+            can_handler.txAddress.can_ifindex = ifr.ifr_ifindex;
+
+            // And bind it to txAddress
+            if (bind(can_handler.socket, (struct sockaddr *)&can_handler.txAddress, sizeof(can_handler.txAddress)) < 0)
+            {
+                AFB_ERROR("d");
+            }
+            else {
+                return 0;
+            }
+        }
+        close(can_handler.socket);
+        can_handler.socket = -1;
+    }
+    return -1;
 }
 
 static int open_can_dev()
 {
-	int rc = retry(open_can_dev_helper);
-	if(rc < 0)
-	{
-		AFB_ERROR("Open of interface %s failed. Falling back to simulation mode", CAN_DEV);
-		can_handler.socket = 0;
-		can_handler.simulation = true;
-		can_handler.send_msg = "FAKE CAN FRAME";
-		rc = 0;
-	}
-	return rc;
+    int rc = retry(open_can_dev_helper);
+    if(rc < 0)
+    {
+        AFB_ERROR("Interface %s failed. Falling back to simulation mode", CAN_DEV);
+        can_handler.socket = 0;
+        can_handler.simulation = true;
+        can_handler.send_msg = "FAKE CAN FRAME";
+        rc = 0;
+    }
+    return rc;
 }
 
 // Get original get temperature function from cpp hvacplugin code
 static uint8_t to_can_temp(uint8_t value)
 {
-	int result = ((0xF0 - 0x10) / 15) * (value - 15) + 0x10;
-	if (result < 0x10)
-		result = 0x10;
-	if (result > 0xF0)
-		result = 0xF0;
+    int result = ((0xF0 - 0x10) / 15) * (value - 15) + 0x10;
+    if (result < 0x10)
+        result = 0x10;
+    if (result > 0xF0)
+        result = 0xF0;
 
-	return (uint8_t)result;
+    return (uint8_t)result;
 }
 
 static uint8_t read_temp_left_zone()
 {
-	return hvac_values[0].value;
+    return hvac_values[0].value;
 }
 
 static uint8_t read_temp_right_zone()
 {
-	return hvac_values[1].value;
+    return hvac_values[1].value;
+}
+
+static uint8_t read_temp_left_led()
+{
+    return hvac_values[5].value;
+}
+
+static uint8_t read_temp_right_led()
+{
+    return hvac_values[6].value;
 }
 
 static uint8_t read_temp()
 {
-	return (uint8_t)(((int)read_temp_left_zone() + (int)read_temp_right_zone()) >> 1);
+    return (uint8_t)(((int)read_temp_left_zone() + (int)read_temp_right_zone()) >> 1);
 }
 
 static uint8_t read_fanspeed()
 {
-	return hvac_values[3].value;
+    return hvac_values[3].value;
+}
+
+/* 
+ * @brief Writing to LED for both temperature sliders
+ */
+
+static int temp_write_led()
+{
+    int rc = 0, red_flag, blue_flag, green_flag, red_value, green_value, blue_value;
+    int right_temp;
+    int left_temp;
+
+    // /sys/class/leds/blinkm-3-9-red/brightness
+    FILE* r = fopen(RED, "w");
+    if (r == NULL) {
+        AFB_ERROR("Unable to open RED path for writing");
+        red_flag = 1;
+    }
+
+    // /sys/class/leds/blinkm-3-9-green/brightness
+    FILE* g = fopen(GREEN, "w");
+    if (g == NULL) {
+        AFB_ERROR("Unable to open GREEN path for writing");
+        green_flag = 1;
+    }
+
+    // /sys/class/leds/blinkm-3-9-blue/brightness
+    FILE* b = fopen(BLUE, "w");
+    if (b == NULL) {
+        AFB_ERROR("Unable to open BLUE path for writing");
+        blue_flag = 1;
+    }
+
+    if (red_flag || green_flag || blue_flag)
+    {
+        rc = 1;
+    }
+
+    left_temp = read_temp_left_led() - 15;
+    right_temp = read_temp_right_led() - 15;
+
+    // Calculates average colour value taken from the temperature toggles
+    red_value = (degree_colours[left_temp].rgb[0] + degree_colours[right_temp].rgb[0]) / 2;
+    green_value = (degree_colours[left_temp].rgb[1] + degree_colours[right_temp].rgb[1]) / 2;
+    blue_value = (degree_colours[left_temp].rgb[2] + degree_colours[right_temp].rgb[2]) / 2;
+
+    // Writes to LED file the specific colour
+    fprintf(r, "%d", red_value);
+    fprintf(g, "%d", green_value);
+    fprintf(b, "%d", blue_value);
+    fclose(r);
+    fclose(g);
+    fclose(b);
+    return rc;
+
+}
+
+/*
+ * @brief Get temperature of left toggle in HVAC system
+ *
+ * @param struct afb_req : an afb request structure
+ *
+ */
+static void temp_left_zone_led(struct afb_req request)
+{
+    int i = 5, rc, x, changed;
+    double d;
+    struct json_object *query, *val;
+    uint8_t values[sizeof hvac_values / sizeof *hvac_values];
+    uint8_t saves[sizeof hvac_values / sizeof *hvac_values];
+
+    AFB_WARNING("In temp_left_zone_led.");
+
+    query = afb_req_json(request);
+
+    /* records initial values */
+    AFB_WARNING("Records initial values");
+    values[i] = saves[i] = hvac_values[i].value;
+
+
+    if (json_object_object_get_ex(query, hvac_values[i].name, &val))
+    {
+        AFB_WARNING("Value of values[i] = %d", values[i]);
+        AFB_WARNING("We got it. Tests if it is an int or double.");
+        if (json_object_is_type(val, json_type_int)) {
+            x = json_object_get_int(val);
+            AFB_WARNING("We get an int: %d",x);
+        }
+        else if (json_object_is_type(val, json_type_double)) {
+            d = json_object_get_double(val);
+            x = (int)round(d);
+            AFB_WARNING("We get a double: %f => %d",d,x);
+        }
+        else {
+            afb_req_fail_f(request, "bad-request",
+                "argument '%s' isn't integer or double", hvac_values[i].name);
+            return;
+        }
+        if (x < 0 || x > 255)
+        {
+            afb_req_fail_f(request, "bad-request",
+                "argument '%s' is out of bounds", hvac_values[i].name);
+            return;
+        }
+        if (values[i] != x) {
+            values[i] = (uint8_t)x;
+            changed = 1;
+            AFB_WARNING("%s changed to %d", hvac_values[i].name,x);
+        }
+    }
+    else {
+        AFB_WARNING("%s not found in query!",hvac_values[i].name);
+    }
+
+
+    if (changed) {
+        i = 5; //(int)(sizeof hvac_values / sizeof *hvac_values);
+        hvac_values[i].value = values[i]; // update structure at line 102
+        AFB_WARNING("WRITE_LED: value: %d", hvac_values[i].value);
+        rc = temp_write_led();
+        if (rc >= 0)
+            afb_req_success(request, NULL, NULL);
+        else if (retry(temp_write_led)) {
+            /* restore initial values */
+            i = 5; //(int)(sizeof hvac_values / sizeof *hvac_values);
+            hvac_values[i].value = saves[i];
+            afb_req_fail(request, "error", "I2C error");
+        }
+    }
+}
+
+/*
+ * @brief Get temperature of right toggle in HVAC system
+ *
+ * @param struct afb_req : an afb request structure
+ *
+ */
+static void temp_right_zone_led(struct afb_req request)
+{
+    int i = 6, rc, x, changed;
+    double d;
+    struct json_object *query, *val;
+    uint8_t values[sizeof hvac_values / sizeof *hvac_values];
+    uint8_t saves[sizeof hvac_values / sizeof *hvac_values];
+
+    AFB_WARNING("In temp_right_zone_led.");
+
+    query = afb_req_json(request);
+
+    /* records initial values */
+    AFB_WARNING("Records initial values");
+    values[i] = saves[i] = hvac_values[i].value;
+
+
+    if (json_object_object_get_ex(query, hvac_values[i].name, &val))
+    {
+        AFB_WARNING("Value of values[i] = %d", values[i]);
+        AFB_WARNING("We got it. Tests if it is an int or double.");
+        if (json_object_is_type(val, json_type_int)) {
+            x = json_object_get_int(val);
+            AFB_WARNING("We get an int: %d",x);
+        }
+        else if (json_object_is_type(val, json_type_double)) {
+            d = json_object_get_double(val);
+            x = (int)round(d);
+            AFB_WARNING("We get a double: %f => %d",d,x);
+        }
+        else {
+            afb_req_fail_f(request, "bad-request",
+                "argument '%s' isn't integer or double", hvac_values[i].name);
+            return;
+        }
+        if (x < 0 || x > 255)
+        {
+            afb_req_fail_f(request, "bad-request",
+                "argument '%s' is out of bounds", hvac_values[i].name);
+            return;
+        }
+        if (values[i] != x) {
+            values[i] = (uint8_t)x;
+            changed = 1;
+            AFB_WARNING("%s changed to %d", hvac_values[i].name,x);
+        }
+    }
+    else {
+        AFB_WARNING("%s not found in query!", hvac_values[i].name);
+    }
+
+
+    if (changed) {
+        i = 6; //(int)(sizeof hvac_values / sizeof *hvac_values);
+        hvac_values[i].value = values[i]; // update structure at line 102
+        AFB_WARNING("WRITE_LED: value: %d", hvac_values[i].value);
+        rc = temp_write_led();
+        if (rc >= 0)
+            afb_req_success(request, NULL, NULL);
+        else if (retry(temp_write_led)) {
+            /* restore initial values */
+            i = 6; //(int)(sizeof hvac_values / sizeof *hvac_values);
+            hvac_values[i].value = saves[i];
+            afb_req_fail(request, "error", "I2C error");
+        }
+    }
 }
 
 static int write_can()
 {
-	struct can_frame txCanFrame;
-	int rc = 0;
-
-	rc = can_handler.socket;
-	if (rc >= 0)
-	{
-		// Hardcoded can_id and dlc (data lenght code)
-		txCanFrame.can_id = 0x30;
-		txCanFrame.can_dlc = 8;
-		txCanFrame.data[0] = to_can_temp(read_temp_left_zone());
-		txCanFrame.data[1] = to_can_temp(read_temp_right_zone());
-		txCanFrame.data[2] = to_can_temp(read_temp());
-		txCanFrame.data[3] = 0xf0;
-		txCanFrame.data[4] = read_fanspeed();
-		txCanFrame.data[5] = 1;
-		txCanFrame.data[6] = 0;
-		txCanFrame.data[7] = 0;
-
-		AFB_DEBUG("%s: %d %d [%02x %02x %02x %02x %02x %02x %02x %02x]\n",
-			can_handler.send_msg,
-			txCanFrame.can_id, txCanFrame.can_dlc,
-			txCanFrame.data[0], txCanFrame.data[1], txCanFrame.data[2], txCanFrame.data[3],
-			txCanFrame.data[4], txCanFrame.data[5], txCanFrame.data[6], txCanFrame.data[7]);
-
-		if(!can_handler.simulation)
-		{
-			rc = (int)sendto(can_handler.socket, &txCanFrame, sizeof(struct can_frame), 0,
-				    (struct sockaddr*)&can_handler.txAddress, sizeof(can_handler.txAddress));
-			if (rc < 0)
-			{
-				AFB_ERROR("Sending CAN frame failed.");
-			}
-		}
-	}
-	else
-	{
-		AFB_ERROR("socket not initialized. Attempt to reopen can device socket.");
-		open_can_dev();
-	}
-	return rc;
+    struct can_frame txCanFrame;
+    int rc = 0;
+
+    rc = can_handler.socket;
+    if (rc >= 0)
+    {
+        // Hardcoded can_id and dlc (data lenght code)
+        txCanFrame.can_id = 0x30;
+        txCanFrame.can_dlc = 8;
+        txCanFrame.data[0] = to_can_temp(read_temp_left_zone());
+        txCanFrame.data[1] = to_can_temp(read_temp_right_zone());
+        txCanFrame.data[2] = to_can_temp(read_temp());
+        txCanFrame.data[3] = 0xf0;
+        txCanFrame.data[4] = read_fanspeed();
+        txCanFrame.data[5] = 1;
+        txCanFrame.data[6] = 0;
+        txCanFrame.data[7] = 0;
+
+        AFB_DEBUG("%s: %d %d [%02x %02x %02x %02x %02x %02x %02x %02x]\n",
+            can_handler.send_msg,
+            txCanFrame.can_id, txCanFrame.can_dlc,
+            txCanFrame.data[0], txCanFrame.data[1], txCanFrame.data[2], txCanFrame.data[3],
+            txCanFrame.data[4], txCanFrame.data[5], txCanFrame.data[6], txCanFrame.data[7]);
+
+        if(!can_handler.simulation)
+        {
+            rc = (int)sendto(can_handler.socket, &txCanFrame, sizeof(struct can_frame), 0,
+                    (struct sockaddr*)&can_handler.txAddress, sizeof(can_handler.txAddress));
+            if (rc < 0)
+            {
+                AFB_ERROR("N frame failed.");
+            }
+        }
+    }
+    else
+    {
+        AFB_ERROR("initialized. Attempt to reopen can device socket.");
+        open_can_dev();
+    }
+    return rc;
 }
 
 /*****************************************************************************************/
 /*****************************************************************************************/
-/**											**/
-/**											**/
-/**	   SECTION: BINDING VERBS IMPLEMENTATION					**/
-/**											**/
-/**											**/
+/**                                         **/
+/**                                         **/
+/**    SECTION: BINDING VERBS IMPLEMENTATION                    **/
+/**                                         **/
+/**                                         **/
 /*****************************************************************************************/
 /*****************************************************************************************/
 
@@ -254,13 +497,13 @@ static int write_can()
  */
 static void get_fanspeed(struct afb_req request)
 {
-	json_object *ret_json;
-	uint8_t fanspeed = read_fanspeed();
+    json_object *ret_json;
+    uint8_t fanspeed = read_fanspeed();
 
-	ret_json = json_object_new_object();
-	json_object_object_add(ret_json, "FanSpeed", json_object_new_int(fanspeed));
+    ret_json = json_object_new_object();
+    json_object_object_add(ret_json, "FanSpeed", json_object_new_int(fanspeed));
 
-	afb_req_success(request, ret_json, NULL);
+    afb_req_success(request, ret_json, NULL);
 }
 
 /*
@@ -271,13 +514,13 @@ static void get_fanspeed(struct afb_req request)
  */
 static void get_temp_right_zone(struct afb_req request)
 {
-	json_object *ret_json;
-	uint8_t temp = read_temp_right_zone();
+    json_object *ret_json;
+    uint8_t temp = read_temp_right_zone();
 
-	ret_json = json_object_new_object();
-	json_object_object_add(ret_json, "RightTemperature", json_object_new_int(temp));
+    ret_json = json_object_new_object();
+    json_object_object_add(ret_json, "RightTemperature", json_object_new_int(temp));
 
-	afb_req_success(request, ret_json, NULL);
+    afb_req_success(request, ret_json, NULL);
 }
 
 /*
@@ -288,13 +531,13 @@ static void get_temp_right_zone(struct afb_req request)
  */
 static void get_temp_left_zone(struct afb_req request)
 {
-	json_object *ret_json;
-	uint8_t temp = read_temp_left_zone();
+    json_object *ret_json;
+    uint8_t temp = read_temp_left_zone();
 
-	ret_json = json_object_new_object();
-	json_object_object_add(ret_json, "LeftTemperature", json_object_new_int(temp));
+    ret_json = json_object_new_object();
+    json_object_object_add(ret_json, "LeftTemperature", json_object_new_int(temp));
 
-	afb_req_success(request, ret_json, NULL);
+    afb_req_success(request, ret_json, NULL);
 }
 
 /*
@@ -305,15 +548,15 @@ static void get_temp_left_zone(struct afb_req request)
  */
 static void get(struct afb_req request)
 {
-	AFB_DEBUG("Getting all values");
-	json_object *ret_json;
+    AFB_DEBUG("Getting all values");
+    json_object *ret_json;
 
-	ret_json = json_object_new_object();
-	json_object_object_add(ret_json, "LeftTemperature", json_object_new_int(read_temp_left_zone()));
-	json_object_object_add(ret_json, "RightTemperature", json_object_new_int(read_temp_right_zone()));
-	json_object_object_add(ret_json, "FanSpeed", json_object_new_int(read_fanspeed()));
+    ret_json = json_object_new_object();
+    json_object_object_add(ret_json, "LeftTemperature", json_object_new_int(read_temp_left_zone()));
+    json_object_object_add(ret_json, "RightTemperature", json_object_new_int(read_temp_right_zone()));
+    json_object_object_add(ret_json, "FanSpeed", json_object_new_int(read_fanspeed()));
 
-	afb_req_success(request, ret_json, NULL);
+    afb_req_success(request, ret_json, NULL);
 }
 
 /*
@@ -324,166 +567,182 @@ static void get(struct afb_req request)
  */
 static void set(struct afb_req request)
 {
-	int i, rc, x, changed;
-	double d;
-	struct json_object *query, *val;
-	uint8_t values[sizeof hvac_values / sizeof *hvac_values];
-	uint8_t saves[sizeof hvac_values / sizeof *hvac_values];
-
-	/* records initial values */
-	AFB_DEBUG("Records initial values");
-	i = (int)(sizeof hvac_values / sizeof *hvac_values);
-	while (i) {
-		i--;
-		values[i] = saves[i] = hvac_values[i].value;
-	}
-
-	/* Loop getting arguments */
-	query = afb_req_json(request);
-	changed = 0;
-	i = (int)(sizeof hvac_values / sizeof *hvac_values);
-	AFB_DEBUG("Looping for args. i: %d", i);
-	while (i)
-	{
-		i--;
-		AFB_DEBUG("Searching... query: %s, i: %d, comp: %s", json_object_to_json_string(query), i, hvac_values[i].name);
-		if (json_object_object_get_ex(query, hvac_values[i].name, &val))
-		{
-			AFB_DEBUG("We got it. Tests if it is an int or double.");
-			if (json_object_is_type(val, json_type_int)) {
-				x = json_object_get_int(val);
-				AFB_DEBUG("We get an int: %d",x);
-			}
-			else if (json_object_is_type(val, json_type_double)) {
-				d = json_object_get_double(val);
-				x = (int)round(d);
-				AFB_DEBUG("We get a double: %f => %d",d,x);
-			}
-			else {
-				afb_req_fail_f(request, "bad-request",
-					"argument '%s' isn't integer or double", hvac_values[i].name);
-				return;
-			}
-			if (x < 0 || x > 255)
-			{
-				afb_req_fail_f(request, "bad-request",
-					"argument '%s' is out of bounds", hvac_values[i].name);
-				return;
-			}
-			if (values[i] != x) {
-				values[i] = (uint8_t)x;
-				changed = 1;
-				AFB_DEBUG("%s changed to %d",hvac_values[i].name,x);
-			}
-		}
-		else {
-			AFB_DEBUG("%s not found in query!",hvac_values[i].name);
-		}
-	}
-
-	/* attemps to set new values */
-	AFB_DEBUG("Diff: %d", changed);
-	if (changed)
-	{
-		i = (int)(sizeof hvac_values / sizeof *hvac_values);
-		while (i) {
-			i--;
-			hvac_values[i].value = values[i];
-		}
-		rc = write_can();
-		if (rc >= 0)
-			afb_req_success(request, NULL, NULL);
-		else if (retry(write_can)) {
-			/* restore initial values */
-			i = (int)(sizeof hvac_values / sizeof *hvac_values);
-			while (i) {
-				i--;
-				hvac_values[i].value = saves[i];
-			}
-			afb_req_fail(request, "error", "CAN error");
-		}
-	}
-	else {
-		afb_req_success(request, NULL, "No changes");
-	}
+    int i, rc, x, changed;
+    double d;
+    struct json_object *query, *val;
+    uint8_t values[sizeof hvac_values / sizeof *hvac_values];
+    uint8_t saves[sizeof hvac_values / sizeof *hvac_values];
+
+    /* records initial values */
+    AFB_DEBUG("Records initial values");
+    i = (int)(sizeof hvac_values / sizeof *hvac_values);
+    while (i) {
+        i--;
+        values[i] = saves[i] = hvac_values[i].value;
+    }
+
+    /* Loop getting arguments */
+    query = afb_req_json(request);
+    changed = 0;
+    i = (int)(sizeof hvac_values / sizeof *hvac_values);
+    AFB_DEBUG("Looping for args. i: %d", i);
+    while (i)
+    {
+        i--;
+        AFB_DEBUG("Searching... query: %s, i: %d, comp: %s", json_object_to_json_string(query), i, hvac_values[i].name);
+        if (json_object_object_get_ex(query, hvac_values[i].name, &val))
+        {
+            AFB_DEBUG("We got it. Tests if it is an int or double.");
+            if (json_object_is_type(val, json_type_int)) {
+                x = json_object_get_int(val);
+                AFB_DEBUG("We get an int: %d",x);
+            }
+            else if (json_object_is_type(val, json_type_double)) {
+                d = json_object_get_double(val);
+                x = (int)round(d);
+                AFB_DEBUG("We get a double: %f => %d",d,x);
+            }
+            else {
+                afb_req_fail_f(request, "bad-request",
+                    "argument '%s' isn't integer or double", hvac_values[i].name);
+                return;
+            }
+            if (x < 0 || x > 255)
+            {
+                afb_req_fail_f(request, "bad-request",
+                    "argument '%s' is out of bounds", hvac_values[i].name);
+                return;
+            }
+            if (values[i] != x) {
+                values[i] = (uint8_t)x;
+                changed = 1;
+                AFB_DEBUG("%s changed to %d",hvac_values[i].name,x);
+            }
+        }
+        else {
+            AFB_DEBUG("%s not found in query!",hvac_values[i].name);
+        }
+    }
+
+    /* attemps to set new values */
+    AFB_DEBUG("Diff: %d", changed);
+    if (changed)
+    {
+        i = (int)(sizeof hvac_values / sizeof *hvac_values);
+        while (i) {
+            i--;
+            hvac_values[i].value = values[i];
+        }
+        rc = write_can();
+        if (rc >= 0)
+            afb_req_success(request, NULL, NULL);
+        else if (retry(write_can)) {
+            /* restore initial values */
+            i = (int)(sizeof hvac_values / sizeof *hvac_values);
+            while (i) {
+                i--;
+                hvac_values[i].value = saves[i];
+            }
+            afb_req_fail(request, "error", "CAN error");
+        }
+    }
+    else {
+        afb_req_success(request, NULL, "No changes");
+    }
 }
 
 int bindingServicePreInit(struct afb_service service)
 {
-	return open_can_dev();
+    return open_can_dev();
 }
 
 int bindingServiceInit(struct afb_service service)
 {
-	event = afb_daemon_make_event("language");
-	if(afb_daemon_require_api("identity", 1))
-		return -1;
-	return afb_service_call_sync("identity", "subscribe", NULL, NULL);
+    event = afb_daemon_make_event("language");
+    if(afb_daemon_require_api("identity", 1))
+        return -1;
+    return afb_service_call_sync("identity", "subscribe", NULL, NULL);
 }
 
 void onEvent(const char *event_name, struct json_object *object)
 {
-	json_object *args, *language = json_object_new_object();
-	json_object *id_evt_name, *current_identity;
-
-	AFB_NOTICE("Event '%s' received: %s", event_name,
-		json_object_to_json_string_ext(object, JSON_C_TO_STRING_PRETTY));
-
-	if (json_object_object_get_ex(object, "eventName", &id_evt_name) &&
-	  !strcmp(json_object_get_string(id_evt_name), "login") &&
-	  !afb_service_call_sync("identity", "get", NULL, &current_identity)) {
-		json_object *response;
-		if (! json_object_object_get_ex(current_identity, "response", &response) || ! json_object_object_get_ex(response, "graphPreferredLanguage", &language)) {
-			language = json_object_new_string("en_US");
-		}
-		afb_event_broadcast(event, language);
-	}
+    json_object *language = json_object_new_object();
+    json_object *id_evt_name, *current_identity;
+
+    AFB_NOTICE("Event '%s' received: %s", event_name,
+        json_object_to_json_string_ext(object, JSON_C_TO_STRING_PRETTY));
+
+    if (json_object_object_get_ex(object, "eventName", &id_evt_name) &&
+      !strcmp(json_object_get_string(id_evt_name), "login") &&
+      !afb_service_call_sync("identity", "get", NULL, &current_identity)) {
+        json_object *response;
+        if (! json_object_object_get_ex(current_identity, "response", &response) || ! json_object_object_get_ex(response, "graphPreferredLanguage", &language)) {
+            language = json_object_new_string("en_US");
+        }
+        afb_event_broadcast(event, language);
+    }
 }
 
 // TODO: Have to change session management flag to AFB_SESSION_CHECK to use token auth
 static const struct afb_verb_v2 _afb_verbs_v2_hvac[]= {
-	{
-		.verb = "get_temp_left_zone",
-		.callback = get_temp_left_zone,
-		.auth = NULL,
-		.info = "Get the left zone temperature",
-		.session = AFB_SESSION_NONE_V2
-	},
-	{
-		.verb = "get_temp_right_zone",
-		.callback = get_temp_right_zone,
-		.auth = NULL,
-		.info = "Get the right zone temperature",
-		.session = AFB_SESSION_NONE_V2
-	},
-	{
-		.verb = "get_fanspeed",
-		.callback = get_fanspeed,
-		.auth = NULL,
-		.info = "Read fan speed",
-		.session = AFB_SESSION_NONE_V2
-	},
-	{
-		.verb = "get",
-		.callback = get,
-		.auth = NULL,
-		.info = "Read all speed",
-		.session = AFB_SESSION_NONE_V2
-	},
-	{
-		.verb = "set",
-		.callback = set,
-		.auth = NULL,
-		.info = "Set a HVAC component value",
-		.session = AFB_SESSION_NONE_V2
-	},
-	{
-		.verb = NULL,
-		.callback = NULL,
-		.auth = NULL,
-		.info = NULL,
-		.session = 0
-	}
+    {
+        .verb = "get_temp_left_zone",
+        .callback = get_temp_left_zone,
+        .auth = NULL,
+        .info = "Get the left zone temperature",
+        .session = AFB_SESSION_NONE_V2
+    },
+    {
+        .verb = "get_temp_right_zone",
+        .callback = get_temp_right_zone,
+        .auth = NULL,
+        .info = "Get the right zone temperature",
+        .session = AFB_SESSION_NONE_V2
+    },
+    {
+        .verb = "get_fanspeed",
+        .callback = get_fanspeed,
+        .auth = NULL,
+        .info = "Read fan speed",
+        .session = AFB_SESSION_NONE_V2
+    },
+    {
+        .verb = "get",
+        .callback = get,
+        .auth = NULL,
+        .info = "Read all speed",
+        .session = AFB_SESSION_NONE_V2
+    },
+    {
+        .verb = "set",
+        .callback = set,
+        .auth = NULL,
+        .info = "Set a HVAC component value",
+        .session = AFB_SESSION_NONE_V2
+    },
+    {
+        .verb = "temp_left_zone_led",
+        .callback = temp_left_zone_led,
+        .auth = NULL,
+        .info = "Turn on LED on left temperature zone",
+        .session = AFB_SESSION_NONE_V2
+
+    },
+        {
+        .verb = "temp_right_zone_led",
+        .callback = temp_right_zone_led,
+        .auth = NULL,
+        .info = "Turn on LED on left temperature zone",
+        .session = AFB_SESSION_NONE_V2
+
+    },
+    {
+        .verb = NULL,
+        .callback = NULL,
+        .auth = NULL,
+        .info = NULL,
+        .session = 0
+    }
 };
 
 const struct afb_binding_v2 afbBindingV2 = {
@@ -495,4 +754,4 @@ const struct afb_binding_v2 afbBindingV2 = {
     .init = bindingServiceInit,
     .onevent = onEvent,
     .noconcurrency = 0
-};
+};
\ No newline at end of file
-- 
2.9.2

